luals-addon/factorio/factorio-plugin/global.lua

```diff
@@ -52,33 +52,39 @@ local function replace(uri, text, diffs)
     this_mod = this_mod:gsub("[^a-zA-Z0-9_]","_")
     local global_name = "__"..this_mod.."__global"
 
-    ---@type {[integer]: integer}
-    local global_matches = {}
-    for start, finish in text:gmatch("%f[a-zA-Z0-9_]()global()%s*[=.%[]")--[[@as fun():integer, integer]] do
-      global_matches[start] = finish
-    end
+    ---@type table<integer, true>
+    local matches_to_ignore = {}
     -- remove matches that where `global` is actually indexing into something (`.global`)
     for dot_pos, start in text:gmatch("()%.[^%S\n]*()global%s*[=.%[]")--[[@as fun():integer, integer]] do
       if text:sub(dot_pos - 1, dot_pos - 1) ~= "." then
-        global_matches[start] = nil
+        matches_to_ignore[start] = true
       end
     end
-    -- `_ENV.global` and `_G.global` now get removed because of this, we can add them back in
```

Comment:
unrelated to the current issue: couldn't you look just before the dotted match just found and skipped for the parent instead of re-doing the whole search, to get _ENV and _G back included?

Reply:
Oh yea absolutely, good idea. I'll just include that in this PR even though it's unrelated

Reply:
It just won't be able to catch them if there's a space after `_G` or `_ENV` (which for some reason the commented out code doesn't check for either), but I think that's still fine, considering 99% of the time there won't be a space.